src/DoocsAdapter.cc

```diff
@@ -73,4 +93,144 @@ namespace ChimeraTK {
+    ChimeraTK::ApplicationBase::getInstance().optimiseUnmappedVariables(pvNames);
+
+    // prepare list of properties connected to the same writable PV
+    {
+      // create map of PV name to properties using the PV (only for writable PVs)
+      std::map<std::string, std::set<std::shared_ptr<ChimeraTK::PropertyDescription>>> reverseMapping;
+      for(const auto& descr : ChimeraTK::VariableMapper::getInstance().getAllProperties()) {
+        for(const auto& source : descr->getSources()) {
+          if(!doocsAdapter.getControlSystemPVManager()->getProcessVariable(source)->isWriteable()) {
+            // to not add PVs to the mapping which are not writeable
+            continue;
+          }
+          reverseMapping[source].insert(descr);
+        }
+      }
+      // filter the map to contain only PVs being used at least twice with at least one writable property
+      for(const auto& p : reverseMapping) {
+        if(p.second.size() < 2) {
+          continue;
+        }
+        size_t writeableCount = 0;
+        for(const auto& d : p.second) {
+          auto attr = std::dynamic_pointer_cast<ChimeraTK::PropertyAttributes>(d);
+          if(attr->isWriteable) ++writeableCount;
+        }
+        if(writeableCount == 0) {
+          continue;
+        }
+        if(writeableCount > 1) {
+          // This case is not (yet) covered. It would require some synchronisation mechanism between writeable
+          // properties, which is not trivial as inconsistencies and dead locks must be avoided.
+          std::cout << "**** WARNING: Variable '" + p.first +
+                  "' mapped to more than one writeable property. Expect inconsistencies!\n";
+        }
+
+        // Add PVs which are used at least twice with at least one writable property to list
+        doocsAdapter.writeableVariablesWithMultipleProperties[p.first] = {};
+      }
+    }
+  }
+
+  /*******************************************************************************************************************/
+
+  /* post_init_epilog is called after all DOOCS properties are fully intialised,
+   * including any value intialisation from the config file. We start the
+   * application here. It will be launched in a separate thread. */
+  void DoocsAdapter::post_init_epilog() {
+    // check for locations not yet created (due to missing entries in the .conf file) and create them now
+    std::map<std::string, int> locMap = ChimeraTK::VariableMapper::getInstance().getLocationAndCode();
+    for(auto& loc : ChimeraTK::VariableMapper::getInstance().getAllLocations()) {
+      int codeToSet = 10;
+      bool defaultUsed = true;
+      // if no code is set in xml file, 10 is default
+      if(locMap.find(loc) != locMap.end()) {
+        codeToSet = locMap.find(loc)->second;
+        defaultUsed = false;
+      }
+
+      auto eq = find_device(loc);
+      if(eq == nullptr) {
+        add_location(codeToSet, loc);
+      }
+      else if(eq->fct_code() != codeToSet && !defaultUsed) {
```

Comment:
This would only be possible if we severely limit the range of the allowed codes, since we would have to instantiate the template for each code "explicitly".